File: scratch/test_aoa.js
const prefabData = [
  { id: 1, tarea: 'EST', desc: 'EREGIR LA ESTRUCTURA', prec: '2', te: 5, cost: 1000 },
  { id: 2, tarea: 'CIM', desc: 'HACER LOS CIMIENTOS', prec: '-', te: 3, cost: 2000 },
  { id: 3, tarea: 'VITE', desc: 'PONER LAS VIGAS TECHO', prec: '1', te: 2, cost: 800 },
  { id: 4, tarea: 'RETE', desc: 'REVESTIR EL TECHO', prec: '3', te: 3, cost: 1200 },
  { id: 5, tarea: 'ELEC', desc: 'CABLEADO ELECTRICO', prec: '1', te: 4, cost: 1500 },
  { id: 6, tarea: 'EXT', desc: 'TABLAS PAREDES EXTERIORES', prec: '7', te: 4, cost: 1100 },
  { id: 7, tarea: 'VENT', desc: 'COLOCAR LAS VENTANAS', prec: '1', te: 2, cost: 500 },
  { id: 8, tarea: 'INT', desc: 'TABLAS PAREDES INTERIORES', prec: '5 ; 7', te: 3, cost: 600 },
  { id: 9, tarea: 'PINT', desc: 'PINTURA EXT. E INT.', prec: '4 ; 6 ; 8', te: 2, cost: 900 }
];

function computeAOA(tasks) {
  const taskMap = {};
  tasks.forEach(t => {
    taskMap[t.tarea] = { ...t, preds: [], succs: [] };
  });

  for (const t of tasks) {
    if (t.prec && t.prec !== '-') {
      const pArr = t.prec.split(';');
      for (const p of pArr) {
        const parent = tasks.find(x => String(x.id) === p.trim() || x.tarea === p.trim());
        if (!parent) return { error: `Precedente no encontrado: ${p}` };
        taskMap[t.tarea].preds.push(parent.tarea);
        taskMap[parent.tarea].succs.push(t.tarea);
      }
    }
  }

  const taskList = Object.values(taskMap);

  const uniqueSets = new Set();
  taskList.forEach(t => {
    const key = t.preds.slice().sort().join(';');
    uniqueSets.add(key);
  });

  taskList.forEach(t => {
    if (t.succs.length > 0) {
      uniqueSets.add(t.tarea);
    }
  });

  const predSetsArray = Array.from(uniqueSets).map(key => {
    return key === '' ? [] : key.split(';');
  });

  let nodeIdCounter = 1;
  const setNodeMap = {};

  function getSetKey(arr) {
    return arr.slice().sort().join(';');
  }

  const emptyKey = getSetKey([]);
  setNodeMap[emptyKey] = nodeIdCounter++;

  predSetsArray.forEach(set => {
    const key = getSetKey(set);
    if (!setNodeMap[key]) {
      setNodeMap[key] = nodeIdCounter++;
    }
  });

  const finalNodeId = nodeIdCounter++;
  let edges = [];

  taskList.forEach(t => {
    const startKey = getSetKey(t.preds);
    const startNode = setNodeMap[startKey];
    
    let endNode;
    if (t.succs.length === 0) {
      endNode = finalNodeId;
    } else {
      const endKey = getSetKey([t.tarea]);
      endNode = setNodeMap[endKey];
    }

    edges.push({
      id: t.tarea,
      tarea: t.tarea,
      desc: t.desc,
      cost: t.cost,
      te: t.te,
      from: startNode,
      to: endNode,
      isDummy: false
    });
  });

  const nonStartSets = predSetsArray.filter(s => s.length > 0);
  
  function isSubset(a, b) {
    return a.every(x => b.includes(x));
  }

  nonStartSets.forEach(B => {
    const keyB = getSetKey(B);
    const nodeB = setNodeMap[keyB];

    const properSubsets = nonStartSets.filter(A => {
      return A.length < B.length && isSubset(A, B);
    });

    const maximalSubsets = properSubsets.filter(A => {
      return !properSubsets.some(C => {
        return C.length > A.length && isSubset(A, C) && isSubset(C, B);
      });
    });

    maximalSubsets.forEach(A => {
      const keyA = getSetKey(A);
      const nodeA = setNodeMap[keyA];
      edges.push({
        id: `Ficticia_${nodeA}_${nodeB}`,
        tarea: 'Ficticia',
        desc: 'Tarea Ficticia',
        cost: 0,
        te: 0,
        from: nodeA,
        to: nodeB,
        isDummy: true
      });
    });

    const covered = new Set();
    maximalSubsets.forEach(A => A.forEach(x => covered.add(x)));
    B.forEach(x => {
      if (!covered.has(x)) {
        const keyA = getSetKey([x]);
        const nodeA = setNodeMap[keyA];
        if (nodeA !== nodeB) {
          edges.push({
            id: `Ficticia_${nodeA}_${nodeB}`,
            tarea: 'Ficticia',
            desc: 'Tarea Ficticia',
            cost: 0,
            te: 0,
            from: nodeA,
            to: nodeB,
            isDummy: true
          });
        }
      }
    });
  });

  let nodes = Array.from(new Set(edges.flatMap(e => [e.from, e.to])));

  let changed = true;
  while (changed) {
    changed = false;
    for (let i = 0; i < edges.length; i++) {
      const e = edges[i];
      if (e.isDummy) {
        if (canContract(i, edges)) {
          contract(i, edges);
          changed = true;
          break;
        }
      }
    }
  }

  nodes = Array.from(new Set(edges.flatMap(e => [e.from, e.to])));

  function hasPath(from, to, edgesList, ignoreIdx) {
    const adj = {};
    edgesList.forEach((edge, idx) => {
      if (idx === ignoreIdx) return;
      if (!adj[edge.from]) adj[edge.from] = [];
      adj[edge.from].push(edge.to);
    });

    const visited = new Set();
    const queue = [from];
    visited.add(from);

    while (queue.length > 0) {
      const curr = queue.shift();
      if (curr === to) return true;
      const neighbors = adj[curr] || [];
      for (const n of neighbors) {
        if (!visited.has(n)) {
          visited.add(n);
          queue.push(n);
        }
      }
    }
    return false;
  }

  function canContract(eIdx, edgesList) {
    const edge = edgesList[eIdx];
    const u = edge.from;
    const v = edge.to;

    if (hasPath(u, v, edgesList, eIdx)) {
      return false;
    }

    const temp = edgesList.map((edge, idx) => {
      if (idx === eIdx) return null;
      const from = edge.from === u ? v : edge.from;
      const to = edge.to === u ? v : edge.to;
      return { from, to };
    }).filter(Boolean);

    for (let i = 0; i < temp.length; i++) {
      for (let j = i + 1; j < temp.length; j++) {
        if (temp[i].from === temp[j].from && temp[i].to === temp[j].to) {
          return false;
        }
      }
    }

    return true;
  }

  function contract(eIdx, edgesList) {
    const edge = edgesList[eIdx];
    const u = edge.from;
    const v = edge.to;

    edgesList.forEach(edge => {
      if (edge.from === u) edge.from = v;
      if (edge.to === u) edge.to = v;
    });

    edgesList.splice(eIdx, 1);
  }

  const adj = {};
  const inDegree = {};
  nodes.forEach(n => {
    adj[n] = [];
    inDegree[n] = 0;
  });
  edges.forEach(e => {
    adj[e.from].push(e.to);
    inDegree[e.to]++;
  });

  const queue = [];
  nodes.forEach(n => {
    if (inDegree[n] === 0) queue.push(n);
  });

  const sortedNodes = [];
  while (queue.length > 0) {
    const curr = queue.shift();
    sortedNodes.push(curr);
    const neighbors = adj[curr] || [];
    for (const n of neighbors) {
      inDegree[n]--;
      if (inDegree[n] === 0) {
        queue.push(n);
      }
    }
  }

  if (sortedNodes.length !== nodes.length) {
    return { error: 'Inconsistencia lógica: Se detectó un ciclo en la red de actividades.' };
  }

  const nodeRenumberMap = {};
  sortedNodes.forEach((oldId, idx) => {
    nodeRenumberMap[oldId] = idx + 1;
  });

  edges.forEach(e => {
    e.from = nodeRenumberMap[e.from];
    e.to = nodeRenumberMap[e.to];
  });

  const N = sortedNodes.length;
  const Fti = Array(N + 1).fill(0);
  const FTj = Array(N + 1).fill(0);

  for (let i = 1; i <= N; i++) {
    const incoming = edges.filter(e => e.to === i);
    if (incoming.length > 0) {
      let maxVal = 0;
      incoming.forEach(e => {
        const val = Fti[e.from] + e.te;
        if (val > maxVal) maxVal = val;
      });
      Fti[i] = maxVal;
    }
  }

  const projectDuration = Fti[N];

  FTj[N] = projectDuration;
  for (let i = N - 1; i >= 1; i--) {
    const outgoing = edges.filter(e => e.from === i);
    let minVal = projectDuration;
    outgoing.forEach(e => {
      const val = FTj[e.to] - e.te;
      if (val < minVal) minVal = val;
    });
    FTj[i] = minVal;
  }

  edges.forEach(e => {
    e.ES = Fti[e.from];
    e.EF = Fti[e.from] + e.te;
    e.LF = FTj[e.to];
    e.LS = FTj[e.to] - e.te;
    e.MT = Math.round((e.LF - e.ES - e.te) * 1000000) / 1000000;
    e.critical = (e.MT === 0);
  });

  const finalNodes = [];
  for (let i = 1; i <= N; i++) {
    finalNodes.push({
      id: i,
      Fti: Fti[i],
      FTj: FTj[i]
    });
  }

  return {
    ok: true,
    nodes: finalNodes,
    edges: edges,
    projectDuration: projectDuration
  };
}

const result = computeAOA(prefabData);
console.log("Project Duration:", result.projectDuration);
console.log("Nodes:", JSON.stringify(result.nodes, null, 2));
console.log("Edges:", JSON.stringify(result.edges.map(e => ({ tarea: e.tarea, from: e.from, to: e.to, te: e.te, ES: e.ES, LF: e.LF, MT: e.MT, critical: e.critical })), null, 2));

File: scratch/test_aoa.py
import json

prefab_data = [
    { "id": 1, "tarea": 'EST', "desc": 'EREGIR LA ESTRUCTURA', "prec": '2', "te": 5, "cost": 1000 },
    { "id": 2, "tarea": 'CIM', "desc": 'HACER LOS CIMIENTOS', "prec": '-', "te": 3, "cost": 2000 },
    { "id": 3, "tarea": 'VITE', "desc": 'PONER LAS VIGAS TECHO', "prec": '1', "te": 2, "cost": 800 },
    { "id": 4, "tarea": 'RETE', "desc": 'REVESTIR EL TECHO', "prec": '3', "te": 3, "cost": 1200 },
    { "id": 5, "tarea": 'ELEC', "desc": 'CABLEADO ELECTRICO', "prec": '1', "te": 4, "cost": 1500 },
    { "id": 6, "tarea": 'EXT', "desc": 'TABLAS PAREDES EXTERIORES', "prec": '7', "te": 4, "cost": 1100 },
    { "id": 7, "tarea": 'VENT', "desc": 'COLOCAR LAS VENTANAS', "prec": '1', "te": 2, "cost": 500 },
    { "id": 8, "tarea": 'INT', "desc": 'TABLAS PAREDES INTERIORES', "prec": '5 ; 7', "te": 3, "cost": 600 },
    { "id": 9, "tarea": 'PINT', "desc": 'PINTURA EXT. E INT.', "prec": '4 ; 6 ; 8', "te": 2, "cost": 900 }
]

def computeAOA(tasks):
    taskMap = {}
    for t in tasks:
        taskMap[t["tarea"]] = {**t, "preds": [], "succs": []}

    for t in tasks:
        if t["prec"] and t["prec"] != '-':
            pArr = [p.strip() for p in t["prec"].split(';') if p.strip()]
            for p in pArr:
                parent = None
                for x in tasks:
                    if str(x["id"]) == p or x["tarea"] == p:
                        parent = x
                        break
                if not parent:
                    return {"error": f"Precedente no encontrado: {p}"}
                taskMap[t["tarea"]]["preds"].append(parent["tarea"])
                taskMap[parent["tarea"]]["succs"].append(t["tarea"])

    taskList = list(taskMap.values())

    uniqueSets = set()
    for t in taskList:
        key = ";".join(sorted(t["preds"]))
        uniqueSets.add(key)

    for t in taskList:
        if len(t["succs"]) > 0:
            uniqueSets.add(t["tarea"])

    predSetsArray = []
    for key in uniqueSets:
        if key == '':
            predSetsArray.append([])
        else:
            predSetsArray.append(key.split(';'))

    nodeIdCounter = 1
    setNodeMap = {}

    def getSetKey(arr):
        return ";".join(sorted(arr))

    emptyKey = getSetKey([])
    setNodeMap[emptyKey] = nodeIdCounter
    nodeIdCounter += 1

    for s in predSetsArray:
        key = getSetKey(s)
        if key not in setNodeMap:
            setNodeMap[key] = nodeIdCounter
            nodeIdCounter += 1

    finalNodeId = nodeIdCounter
    nodeIdCounter += 1
    edges = []

    for t in taskList:
        startKey = getSetKey(t["preds"])
        startNode = setNodeMap[startKey]
        
        if len(t["succs"]) == 0:
            endNode = finalNodeId
        else:
            endKey = getSetKey([t["tarea"]])
            endNode = setNodeMap[endKey]

        edges.append({
            "id": t["tarea"],
            "tarea": t["tarea"],
            "desc": t["desc"],
            "cost": t["cost"],
            "te": t["te"],
            "from": startNode,
            "to": endNode,
            "isDummy": False
        })

    nonStartSets = [s for s in predSetsArray if len(s) > 0]
    
    def isSubset(a, b):
        return all(x in b for x in a)

    for B in nonStartSets:
        keyB = getSetKey(B)
        nodeB = setNodeMap[keyB]

        properSubsets = [A for A in nonStartSets if len(A) < len(B) and isSubset(A, B)]

        maximalSubsets = []
        for A in properSubsets:
            # Check if there is another set C in properSubsets such that A C C C B
            is_max = True
            for C in properSubsets:
                if len(C) > len(A) and isSubset(A, C) and isSubset(C, B):
                    is_max = False
                    break
            if is_max:
                maximalSubsets.append(A)

        for A in maximalSubsets:
            keyA = getSetKey(A)
            nodeA = setNodeMap[keyA]
            edges.append({
                "id": f"Ficticia_{nodeA}_{nodeB}",
                "tarea": "Ficticia",
                "desc": "Tarea Ficticia",
                "cost": 0,
                "te": 0,
                "from": nodeA,
                "to": nodeB,
                "isDummy": True
            })

        covered = set()
        for A in maximalSubsets:
            for x in A:
                covered.add(x)
        for x in B:
            if x not in covered:
                keyA = getSetKey([x])
                nodeA = setNodeMap[keyA]
                if nodeA != nodeB:
                    edges.append({
                        "id": f"Ficticia_{nodeA}_{nodeB}",
                        "tarea": "Ficticia",
                        "desc": "Tarea Ficticia",
                        "cost": 0,
                        "te": 0,
                        "from": nodeA,
                        "to": nodeB,
                        "isDummy": True
                    })

    def hasPath(start, target, edgesList, ignoreIdx):
        adj = {}
        for idx, edge in enumerate(edgesList):
            if idx == ignoreIdx:
                continue
            if edge["from"] not in adj:
                adj[edge["from"]] = []
            adj[edge["from"]].append(edge["to"])

        visited = set()
        queue = [start]
        visited.add(start)

        while queue:
            curr = queue.pop(0)
            if curr == target:
                return True
            neighbors = adj.get(curr, [])
            for n in neighbors:
                if n not in visited:
                    visited.add(n)
                    queue.append(n)
        return False

    def canContract(eIdx, edgesList):
        edge = edgesList[eIdx]
        u = edge["from"]
        v = edge["to"]

        if hasPath(u, v, edgesList, eIdx):
            return False

        temp = []
        for idx, edge_item in enumerate(edgesList):
            if idx == eIdx:
                continue
            fr = v if edge_item["from"] == u else edge_item["from"]
            to = v if edge_item["to"] == u else edge_item["to"]
            temp.append((fr, to))

        for i in range(len(temp)):
            for j in range(i + 1, len(temp)):
                if temp[i][0] == temp[j][0] and temp[i][1] == temp[j][1]:
                    return False
        return True

    def contract(eIdx, edgesList):
        edge = edgesList[eIdx]
        u = edge["from"]
        v = edge["to"]

        for edge_item in edgesList:
            if edge_item["from"] == u:
                edge_item["from"] = v
            if edge_item["to"] == u:
                edge_item["to"] = v

        edgesList.pop(eIdx)

    changed = True
    while changed:
        changed = False
        for i in range(len(edges)):
            e = edges[i]
            if e["isDummy"]:
                if canContract(i, edges):
                    contract(i, edges)
                    changed = True
                    break

    nodes = list(set(e["from"] for e in edges) | set(e["to"] for e in edges))

    adj = {n: [] for n in nodes}
    inDegree = {n: 0 for n in nodes}
    for e in edges:
        adj[e["from"]].append(e["to"])
        inDegree[e["to"]] += 1

    queue = [n for n in nodes if inDegree[n] == 0]
    sortedNodes = []
    while queue:
        curr = queue.pop(0)
        sortedNodes.append(curr)
        for neighbors in adj.get(curr, []):
            inDegree[neighbors] -= 1
            if inDegree[neighbors] == 0:
                queue.append(neighbors)

    nodeRenumberMap = {oldId: idx + 1 for idx, oldId in enumerate(sortedNodes)}

    for e in edges:
        e["from"] = nodeRenumberMap[e["from"]]
        e["to"] = nodeRenumberMap[e["to"]]

    N = len(sortedNodes)
    Fti = [0] * (N + 1)
    FTj = [0] * (N + 1)

    for i in range(1, N + 1):
        incoming = [e for e in edges if e["to"] == i]
        if incoming:
            maxVal = 0
            for e in incoming:
                val = Fti[e["from"]] + e["te"]
                if val > maxVal:
                    maxVal = val
            Fti[i] = maxVal

    projectDuration = Fti[N]

    FTj[N] = projectDuration
    for i in range(N - 1, 0, -1):
        outgoing = [e for e in edges if e["from"] == i]
        minVal = projectDuration
        for e in outgoing:
            val = FTj[e["to"]] - e["te"]
            if val < minVal:
                minVal = val
        FTj[i] = minVal

    for e in edges:
        e["ES"] = Fti[e["from"]]
        e["EF"] = Fti[e["from"]] + e["te"]
        e["LF"] = FTj[e["to"]]
        e["LS"] = FTj[e["to"]] - e["te"]
        e["MT"] = e["LF"] - e["ES"] - e["te"]
        e["critical"] = (e["MT"] == 0)

    finalNodes = [{"id": i, "Fti": Fti[i], "FTj": FTj[i]} for i in range(1, N + 1)]

    return {
        "nodes": finalNodes,
        "edges": edges,
        "projectDuration": projectDuration
    }

res = computeAOA(prefabData)
print("Duration:", res["projectDuration"])
print("Nodes:")
for n in res["nodes"]:
    print(n)
print("Edges:")
for e in res["edges"]:
    print(f"{e['tarea']} ({e['from']} -> {e['to']}), te={e['te']}, ES={e['ES']}, LF={e['LF']}, MT={e['MT']}, critical={e['critical']}")
